website/pages/docs/connect/observability/ui-visualization.mdx

```diff
@@ -0,0 +1,501 @@
+allowlist to avoid exposing unintended parts of the API. For example with
+Prometheus, both the `/api/v1/query_range` and `/api/v1/query` endpoints are
+needed to load time-series and individual stats. If the proxy had the `base_url`
+set to `http://prometheus-server` then the proxy would also expose read access
+to several other endpoints such as `/api/v1/status/config` which includes all
+Prometheus configuration which might include sensitive information.
+
+The default value for `metrics_proxy.path_allowlist` is `["/api/v1/query_range",
+"/api/v1/query"]` so that if you are using the built in provider, this is
+already limited to the essential endpoints.
+
+If you use a custom provider that uses the metrics proxy, you'll need to
+explicitly set the allowlist based on the endpoints the provider needs to
+access.
+
+#### Adding Headers
+
+It is also possible to configure the proxy to add one or more headers to
+requests as they pass through. This is useful when the metrics backend requires
+authentication. For example if your metrics are shipped to a hosted provider,
+you could provision an API token specifically for the Consul UI and configure
+the proxy to add it as in the example below. This keeps the API token only
+visible to Consul operators in the configuration file while UI users can query
+the metrics they need without separately obtaining a token for that provider or
+having a token exposed to them that they might be able to use elsewhere.
+
+```hcl
+ui_config {
+  enabled = true
+  metrics_provider = "example-apm"
+  metrics_proxy {
+    base_url = "http://example-apm/api/v1/metrics"
+    add_headers = [
+      {
+        name = "Authorization"
+        value = "Bearer <token>"
+      }
+    ]
+  }
+}
+```
+
+## Custom Metrics Providers
+
+Consul 1.9.0 includes a built-in provider for fetching metrics from
+[Prometheus](https://prometheus.io). To enable others to use the feature along
+with other existing metrics stores and hosted services, we created a "metrics
+provider" interface in Javascript. A custom provider may be written and the
+javascript file served by the Consul agent.
+
+~> **Note**: this interface is _experimental_ and may change in breaking ways or
+ be removed entirely as we discover the needs of the community. Please provide
+ feedback on Github or [Discuss](https://discuss.hashicorp.com/) on how you'd
+ like to use this.
+
+The template for a complete provider Javascript file is given below.
+
+```Javascript
+(function () {
+  var provider = {
+    /**
+     * init is called when the provider is first loaded.
+     *
+     * options.providerOptions contains any operator configured parameters
+     * specified in the `metrics_provider_options_json` field of the Consul
+     * agent configuration file.
+     *
+     * Consul will provide a boolean field options.metrics_proxy_enabled to
+     * indicate whether the agent has a metrics proxy configured.
```

Comment:
```suggestion
     * Consul will provide:
     *
     * 1. A boolean field options.metrics_proxy_enabled to indicate whether the agent
     *    has a metrics proxy configured.
     *
     * 2. A fetch-like options.fetch which is a thin fetch wrapper that prefixes
     *    any url with the url of Consul's proxy endpoint and adds your current
     *    Consul ACL token to the request headers. Otherwise it functions like the
     *    browsers native fetch
```

Reply:
What does "fetch-like" mean? Shouldn't we just describe the method signature?

Reply:
Oh nm, I read to the end TIL about https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API I'll link those API docs to make this clear to others who've not head about it before.